### PR TITLE
Deprecation function to be mutable in all cases

### PR DESCRIPTION
**Pull request by HelenCampbell**

### CODE REVIEW

File: lib/puppet/parser/functions/deprecation.rb [hunk 1]
```diff
@@ -4,12 +4,15 @@ module Puppet::Parser::Functions
 EOS
   ) do |arguments|
 
+
```
Comment (tphoney) [inline lib/puppet/parser/functions/deprecation.rb]: extra line not needed

File: spec/functions/is_array_spec.rb [hunk 2]
```diff
@@ -21,4 +16,17 @@
   it { is_expected.to run.with_params('one').and_return(false) }
   it { is_expected.to run.with_params(1).and_return(false) }
   it { is_expected.to run.with_params({}).and_return(false) }
+  context 'Checking for deprecation warning' do
```
Comment (tphoney) [inline spec/functions/is_array_spec.rb]: should this be at the bottom ? it being at the top alerts people that this function is about to be deprecated, even in the tests

File: spec/functions/deprecation_spec.rb [hunk 3]
```diff
@@ -42,6 +42,7 @@
   end
 else
   describe 'deprecation' do
+    ENV['STDLIB_LOG_DEPRECATIONS'] = "true"
```
Comment (DavidS) [inline spec/functions/deprecation_spec.rb]: This still needs to go in a `before` [hook](https://www.relishapp.com/rspec/rspec-core/docs/hooks/before-and-after-hooks), and there should be tests for each possible value(-set).

Comment (HelenCampbell) [inline spec/functions/deprecation_spec.rb]: After discussion David and I decided the code change to add before hooks and contexts is too much for the simplicity of the change. Instead I will add after hooks to each of the test for cleanup.

**Pull request merged**